docs/devel/coding-conventions.md

```diff
@@ -30,12 +30,53 @@ Documentation for other releases can be found at
 <!-- END STRIP_FOR_RELEASE -->
 
 <!-- END MUNGE: UNVERSIONED_WARNING -->
-Coding style advice for contributors
+Code conventions
   - Bash
     - https://google-styleguide.googlecode.com/svn/trunk/shell.xml
+    - Ensure that build, release, test, and cluster-management scripts run on OS X
   - Go
-    - https://github.com/golang/go/wiki/CodeReviewComments
-    - https://gist.github.com/lavalamp/4bd23295a9f32706a48f
+    - Ensure your code passes the [presubmit checks](development.md#hooks)
+    - [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
+    - [Effective Go](https://golang.org/doc/effective_go.html)
+    - Command-line flags should use dashes, not underscores
+    - Naming
+      - Please consider package name when selecting an interface name, and avoid redundancy.
+        - e.g.: `storage.Interface` is better than `storage.StorageInterface`.
+      - Do not use uppercase characters, underscores, or dashes in package names.
+      - Please consider parent directory name when choosing a package name.
+        - so pkg/controllers/autoscaler/foo.go should say `package autoscaler` not `package autoscalercontroller`.
+        - Unless there's a good reason, the `package foo` line should match the name of the directory in which the .go file exists.
+        - Importers can use a different name if they need to disambiguate.
+    - API conventions
+      - [API changes](api_changes.md)
+      - [API conventions](api-conventions.md)
+    - [Kubectl conventions](kubectl-conventions.md)
+    - [Logging conventions](logging.md)
+
+Testing conventions
+  - All new packages and most new significant functionality must come with unit tests
+  - Table-driven tests are preferred for testing multiple scenarios/inputs; for example, see [TestNamespaceAuthorization](../../test/integration/auth_test.go)
+  - Significant features should come with integration (test/integration) and/or end-to-end (test/e2e) tests
+    - Including new kubectl commands and major features of existing commands
+  - Unit tests must pass on OS X and Windows platforms - if you use Linux specific features, your test case must either be skipped on windows or compiled out (skipped is better when running Linux specific commands, compiled out is required when your code does not compile on Windows).
+
+Directory and file conventions
+    - Avoid package sprawl. Find an appropriate subdirectory for new packages. (See [#4851](http://issues.k8s.io/4851) for discussion.)
+      - Libraries with no more appropriate home belong in new package subdirectories of pkg/util
+    - Avoid general utility packages. Packages called "util" are suspect. Instead, derive a name that describes your desired function. For example, the utility functions dealing with waiting for operations are in the "wait" package and include functionality like Poll. So the full name is wait.Poll
+    - Directory and document filenames should use dashes, not underscores
```

Comment:
I fixed the wording. I didn't address the issue about stuff other than Go and docs. I imagine that other things have their own conventions/constraints, which we can document as we discover them.